staff/puppet/gen-rootpw

```diff
@@ -1,28 +1,52 @@
-#!/bin/bash -e
-# Update the OCF root password. We use a ridiculous number of hash iterations
-# to make cracking the hashes way over-the-top impossible.
-
-ROOT_PW_FILE="/opt/puppetlabs/shares/private/rootpw"
-
-echo "########################################################"
-echo "This will update the root password for ALL OCF MACHINES."
-echo "    YOU PROBABLY SHOULD NOT BE RUNNING THIS COMMAND."
-echo "########################################################"
-echo
-read -rp "Are you sure you want to do this? [yN] " answer
-
-if [ "$answer" != "y" ]; then
-    echo "Wise choice."
-    exit 1
-fi
-
-HASH=$(mkpasswd -m sha-512 -R 65536 | tr -d '\n')
-
-if [ -z "$HASH" ]; then
-    echo "Didn't get a password?"
-    exit 1
-fi
-
-echo -n "$HASH" > $ROOT_PW_FILE
-chown puppet:puppet $ROOT_PW_FILE
-chmod 600 $ROOT_PW_FILE
+#!/usr/bin/env python3
+import crypt
+import subprocess
+import sys
+from getpass import getpass
+from hmac import compare_digest
+
+
+WARNING_MESSAGE = """\
+################################################################################
+This will generate a new password hash, but not update it across machines
+
+To do that, you will need to put this new password hash into the puppet
+private hiera file. Do NOT do this without input from others!
+################################################################################
+"""
+
+
+def main():
+    print(WARNING_MESSAGE)
+
+    password = getpass('New password: ')
+    password_confirm = getpass('New password confirmation: ')
+
+    if password != password_confirm:
+        print('Password and confirmation do not match, please try again')
+        sys.exit(1)
+
+    hashed_password = subprocess.check_output(
+        ('mkpasswd', '-m', 'sha-512', '-R', '65536', password),
+    ).decode('utf-8').strip()
+
+    # Check that the hash from mkpasswd matches one generated through python as
+    # an extra sanity check
+    hashed_array = hashed_password.split('$')
+    hashed_confirmation = crypt.crypt(
+        password,
+        '${}${}${}'.format(hashed_array[1], hashed_array[2], hashed_array[3]),
+    )
+
+    if not compare_digest(hashed_password, hashed_confirmation):
+        print('Password hash generated by python did not match that generated')
+        print("by mkpasswd, make sure that there aren't any newlines, etc. in")
+        print('your password')
+        sys.exit(2)
+
+    print('Here is the new password hash, go update ocf::root_password in puppet:')
+    print(hashed_password)
+
+
+if __name__ == '__main__':
+    sys.exit(main())
```

Comment:
Very minor nit but you `sys.exit` in `main()` as well as here. It might be nice to choose one, i.e. just `sys.exit(0)` at the end.

Reply:
Ah nice, yeah it should just be a `return` in the main function instead, thanks for noticing that